elastic/beats
# Update Metricbeat, Filebeat, libbeat with elastic-agent V2 support

## What does this PR do?
This PR has a few features:

- It adds support for the V2 elastic agent client controller
- Creates a config shim callback system to replace the V1's AST config transformation, requiring all V2 beats to define a config transformation callback
- Adds a test suite system with a mocked V2 controller that allows for testing for individual beats
- Removes V1 support


## What are the components of this PR?

- `libbeat/cfgfile/cfgfile.go`
This prevents the beats from reading a config file when they start up in remove management mode.

- beat-specific changes `metricbeat/beater/metricbeat.go`, `x-pack/metricbeat/cmd/agent.go`, etc.
These changes register the reloader interfaces, and define beat-specific config transformations that act as a compatiblity shim between the V2 config structure and pre-existing beats config

- `x-pack/libbeat/management/*`
The actual V2 client code that reloads beats and accepts commands from a V2 control server.

- `x-pack/libbeat/management/tests/*`
An integration test suite that initializes an entire beat, creates a mock elastic-agent server, lets the beat write any metrics to a logfile, then verifies the metrics, fields, and configs are expected.

## What's missing in this PR

Due to a combination of time and knowledge constraints, there are a number of things this PR doesn't do with regards to V2 support:

- This PR only adds support for metricbeat and filebeat. Due to the idiosyncrasy of individual beats, support for other beats will need to come from code owners on those beats.
- Elasticsearch header transformations. The V1 AST transformations, now gone in V2, have a `InjectHeadersRule` that will modify the header config of an elasticsearch output. These headers were obtained via an `AgentInfo` struct that currently exists in V2, but it doesn't support a `Headers` field.

## Checklist

## Author's Checklist

- [x] Write a markdown document to help other devs port beat-specific code to V2
- [x] Add more filebeat tests
- [x] Make the V2 client handle sigint properly
- [ ] Test against a "live" agent once this PR is merged: https://github.com/elastic/elastic-agent/pull/850

## How to test this PR locally

Check out the `tests` directory in `x-pack/libbeat/management`. Each sub-directory contains a number of folders with integration tests for each beat. Note that until https://github.com/elastic/elastic-agent/pull/850 is merged, this won't work against a real agent, and the mock server in the test suite will be required to actually test the V2 client code.
